Traders chat

chimp1984 opened the pull request:
Based on PR  #2871 from @sqrrm.

I share the PR so that others can add optimisations for design and wording.

Open questions:
- Do we need a flag in the settings to allow the user to not receive chat messages? Would require that the peer gets that information as well which is a bit of an effort. Maybe we can keep that for a later version if needed at all.
- Do we need a popup or "news" badge to make sure users will find the new feature? I guess it is obvious enough, specially if you get a message.
- Do we want a badge icon showing new messages if the selected screen is not pending trades?
- Do we want to send mobile notifications?

I think all the above is not needed, at least not for a first release. But some UI beautification (@ripcurlx) and better wording (@m52go) would be nice...

<img width="1423" alt="Screen Shot 2019-08-30 at 01 13 20" src="https://user-images.githubusercontent.com/54558767/63982606-6301b300-cac3-11e9-8ebd-1c8c6c131e5c.png">

sqrrm committed:
Refactor Chat out of TraderDisputeView

On the way to adding chat for traders this is a first step. Mainly just
moving functionality out of TraderDisputeView to Chat class. There are
still remaining dispute functionality that needs to be factored away.

sqrrm committed:
Send acks when capability found

sqrrm committed:
Separate most of Chat and arbitration

The naming of DisputeCommunicationMessage has to stay but they otherwise
fit what would be more aptly named ChatCommunicationMessage or something
in that spririt.

sqrrm committed:
Move close ticket button creation to arbitration

Add extra button for generically added button in chat

sqrrm committed:
Handle Chat message subscription inside Chat

sqrrm committed:
Add trader chat

- Add communication messages to Trade protobuf message to be able to
save chat messages per trade
- Add Type enum and field to DisputeCommunicationMessage protobuf to
be able to dispatch Dispute and Trade chat messages properly
- Rename some function as isClient instead of isTrader to make it easier
to understand who is who when two traders are communicating with each
other

sqrrm committed:
Add ChatManager

Move session classes to core. Break out DisputeCommunicationMessage
handling from DisputeManager and put in ChatMananger prepare for other
uses of ChatManager.

Renaming of DisputeCommunicationMessage would be nice but it's
representing the protobuf messages so the name has to stay.

sqrrm committed:
Make attachments optional

sqrrm committed:
Test trade chat UI

sqrrm committed:
Add comments on BSQ trade fee calculation

sqrrm committed:
Change message box aid

sqrrm committed:
Add basic chat UI

Very basic switch between chat and overview per trade, improvements
needed on the UI.

chimp1984 committed:
Merge branch 'master_upstream' into sq-add-trader-chat

# Conflicts:
#	core/src/main/java/bisq/core/arbitration/DisputeManager.java
#	core/src/main/java/bisq/core/arbitration/messages/DisputeCommunicationMessage.java
#	core/src/main/java/bisq/core/trade/Trade.java

chimp1984 committed:
Fix merge issues

chimp1984 committed:
Update string

chimp1984 committed:
Add comments, cleanup, fix wrong param name

chimp1984 committed:
Show chat in popup instead of tab (WIP)

chimp1984 committed:
Add close handler to popup

- Close already open chat if new one opens

chimp1984 committed:
Merge branch 'master_upstream' into sq-add-trader-chat

# Conflicts:
#	desktop/src/main/java/bisq/desktop/main/disputes/trader/TraderDisputeView.java

chimp1984 committed:
Apply changes from master

chimp1984 committed:
Add missing allServicesInitialized check

chimp1984 committed:
Add wasDisplayed field to chatMsg

- Fix logger
- Manage display state off trade chat msg (wip)

chimp1984 committed:
Add onAllServicesInitialized call, rename method

chimp1984 committed:
Add missing setWasDisplayed

chimp1984 committed:
Add badge for num new chat msg

chimp1984 committed:
Fix wrong pubkey

chimp1984 committed:
Add chat rules system msg; cleanups

chimp1984 committed:
Handle nullables

battleofwizards commented:
> I think all the above is not needed, at least not for a first release.

I agree. None of the mentioned improvements sound critical for the first release. Better to release the thinnest vertically complete features possible.

ripcurlx commented:
@chimp1984 Looking great already! I'll give it a spin and provide feedback soon.

ripcurlx committed:
Fix positioning of badge

ripcurlx commented:
I fixed the badge positioning and the tooltip font size. I'd also suggest following changes to the icon: Using a material design version of the chat icon and color it in our action color (green), which is more consistent throughout the app.
![Bildschirmfoto 2019-08-30 um 12 37 25](https://user-images.githubusercontent.com/170962/64014655-f918f600-cb22-11e9-9f31-1721495c912e.png)

ripcurlx committed:
Use button instead of label to fix tooltip font size, use material design icon and add additional cleanup code

ripcurlx commented:
I just did more in-depth testing and everything works as expected. Also having arbitration case in parallel to the trader chat for a specific message doesn't cause any issues. Great work @sqrrm and @chimp1984 👍 
@m52go Could you review the text content if you want to have some changes there?

If that is done I'm happy to merge this PR to master.

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -608,6 +608,18 @@ portfolio.pending.step2_seller.waitPayment.msg=The deposit transaction has at le
 
+tradeChat.chatWindowTitle=Chat window for trade with ID ''{0}''
+tradeChat.openChat=Open chat window
+tradeChat.rules=You can communicate with your trade peer to resolve potential problems with this trade. It is not mandatory to reply in \
```
```
tradeChat.rules=Here you can communicate with your trading peer to resolve potential problems with this trade.\n\n\
It is not mandatory to reply to any messages.\n\nIf a trader violates any of the rules below, open a dispute with \
'Ctrl/Cmd + o' and report it to the arbitrator.\n\n\
```

Main updates here: 

- put "it is not mandatory to reply to any messages" on its own line so it is crystal clear
- Add 'Ctrl/Cmd' for non-Mac users

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -608,6 +608,18 @@ portfolio.pending.step2_seller.waitPayment.msg=The deposit transaction has at le
+tradeChat.rules=You can communicate with your trade peer to resolve potential problems with this trade. It is not mandatory to reply in \
+  the chat. If a trader violates the below rules, open a dispute with 'cmd + o' and report it to the arbitrator.\n\n\
+  Chat rules:\n\
+  \t● Do not send links to unknown URLs (risk of malware). Links to a block explorer are permitted.\n\
```
Should we also expressly disallow link shorteners?

Separately: even bad links can be disguised to look like known URLs with weird TLDs. I think all links should be disallowed, but maybe I'm paranoid.

chimp1984 replied:
Yes, good point with link shortener. I think it should be added. Maybe also recommend to use Tor browser as that is more safe with deactivated js etc...

chimp1984 replied:
Right. Actually beside links to the tx at a block-explorer no links would be needed. And those can be looked up by the other used with the tx id as well. So right maybe disallow any URLs?

m52go replied:
> So right maybe disallow any URLs?

Yes, I would say so, unless anyone has a solid reason to keep them.

```suggestion
  \t● URLs are not allowed. If you would like to send a link to a block explorer, for example, just send the transaction ID, and the other peer can look it up on their own.\n\
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -608,6 +608,18 @@ portfolio.pending.step2_seller.waitPayment.msg=The deposit transaction has at le
+  the chat. If a trader violates the below rules, open a dispute with 'cmd + o' and report it to the arbitrator.\n\n\
+  Chat rules:\n\
+  \t● Do not send links to unknown URLs (risk of malware). Links to a block explorer are permitted.\n\
+  \t● Do not try to motivate peer to trade outside of Bisq (no security).\n\
```
```suggestion
  \t● Do not encourage trading outside of Bisq (no security).\n\
```

m52go commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -608,6 +608,18 @@ portfolio.pending.step2_seller.waitPayment.msg=The deposit transaction has at le
+  \t● Do not try to motivate peer to trade outside of Bisq (no security).\n\
+  \t● Keep conversation friendly and respectful.\n\
+  \t● Do not engage in any form of social engineering scam attempts.\n\
+  \t● Respect if the peer is not responding and prefers to not communicate via the chat.\n\
```
```suggestion
  \t● If a peer is not responding and prefers to not communicate via chat, respect their decision.\n\
```

m52go replied:
I would also add an additional rule, as it comes up quite a bit, but is a huge problem for privacy:

```
\t● Do not attempt to contact the trading peer outside of this chat.\n\
```

wiz replied:
Do not copy and paste your seed words, private keys, passwords, or other sensitive information into this chat!

chimp1984 replied:
> Do not attempt to contact the trading peer outside of this chat

Not so sure if that should be a strict rule as in cases of network instability users do communicate via email if the email is known. I would say lets leave that to common sense of users...

chimp1984 replied:
Updated version:

```
tradeChat.rules=You can communicate with your trade peer to resolve potential problems with this trade.\n\
  It is not mandatory to reply in the chat.\n\
  If a trader violates the below rules, open a dispute with 'Cmd/Ctrl + o' and report it to the arbitrator.\n\n\
  Chat rules:\n\
  \t● Do not send any links (risk of malware). You can send the transaction ID and the name of a block explorer.\n\
  \t● Do not send your seed words, private keys, passwords or other sensitive information!\n\
  \t● Do not encourage trading outside of Bisq (no security).\n\
  \t● Do not engage in any form of social engineering scam attempts.\n\
  \t● If a peer is not responding and prefers to not communicate via chat, respect their decision.\n\
  \t● Keep conversation scope limited to the trade. This chat is not a messenger replacement or trollbox.\n\
  \t● Keep conversation friendly and respectful.
```

m52go reviewed (commented):
Text looks good to me overall...just a potential issue with links (see comment).

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update core/src/main/resources/i18n/displayStrings.properties

Co-Authored-By: Steve Jain <mfiver@gmail.com>

chimp1984 committed:
Update tradeChat.rules

chimp1984 committed:
Persist chat position. Improve listener handling.

- Store last position of the chat window so if it gets closed and opened
again it opens at the last position.

- Fix issues with the listener for new messages.
The handler was called multiple times before. Now its is called only
once. Tested with multiple trades and scrolling.

We use maps for each trade to avoid multiple listener registrations
when switching views. With current implementation we avoid that but we
do not remove listeners when a trade is removed (completed) but that
has no consequences as we will not receive any message anyway from a
closed trade. Supporting it more correctly would require more effort
and managing listener deactivation at screen switches (currently we
get the update called if we have selected another view.

This part can be improved if any dev feels motivated but its not
trivial...

chimp1984 committed:
Close chat window if dispute gets closed

ripcurlx reviewed (approved):
ACK

sqrrm commented on desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesView.java:
```diff
@@ -316,6 +316,19 @@ private void openChat(Trade trade) {
         if (chatPopupStage != null)
             chatPopupStage.close();
 
+        if (trade.getCommunicationMessages().isEmpty()) {
```
This will make receivers of messages from peers not see the code of conduct message
